### DISCUSSION TITLE
인터페이스의 디폴트 메서드

### DISCUSSION THREAD

**Discussion by 0sunset0** [cowaine/is-my-code-that-strange#56] [Chapter 06 (조건 분기)] [answered]
코드 6.28에는 Magic 인터페이스가 있고, 여러 클래스들이 이를 구현하고 있습니다. 
```java
public interface Magic {

    String name();
    MagicPoint costMagicPoint();
    AttackPower attackPower();
    TechnicalPoint costTechnicalPoint();
}
```

<br>
이때 만약 인터페이스에 `전역적으로 사용될 새로운 기능`이 추가되어야 한다면, 
디폴트 메서드를 사용해야 할까요 아니면 추상메서드를 선언해 구현클래스에서 구현해야 할까요? 

```java
public interface Magic {

    String name();
    MagicPoint costMagicPoint();
    AttackPower attackPower();
    TechnicalPoint costTechnicalPoint();

    // 디폴트 메서드
    default String getMagicClassName() {
        return this.getClass().getSimpleName();
    }
}
```

```java
public interface Magic {

    String name();
    MagicPoint costMagicPoint();
    AttackPower attackPower();
    TechnicalPoint costTechnicalPoint();

    //추상 메서드
    String getMagicClassName();
}
```

디폴트 메서드의 장단점을 설명해주세요! 그리고 인터페이스 디폴트 메서드를 만들어보신 적 있는지 궁금합니다 🤔

**Comment by firewoody237** [comment]
모든 `Magic`을 구현하는 클래스들이 같은 결과를 내야 하는 메서드라면 `default`로,
각자 다른 역할을 하는 메서드라면 각 클래스에서 구현하겠습니다.

OCP를 지킬 수 있는게 가장 큰 장점이지 않을까 싶네요.
단점은 용도에 맞춰서 쓴다면 딱히 생각이 안나는거 같습니다.

**Comment by corock** [comment, marked as answer]
상속은 객체지향의 특징 중 하나로 항상 소개되지만, 부모와 자식 클래스 간 결합도가 높아져서 저는 상속 사용을 최대한 자제하고 있습니다.
'전역적으로 사용될 새로운 기능' 이 무엇인지에 따라 구현 방법은 다를 것 같은데 디폴트 메서드(_default method_)를 선호할 것 같습니다.

_default method_ 사용 경험은 DTO 와 엔티티를 변환할 때 `toDto` 나 `toEntity` 메서드 등을 서비스 인터페이스에 구현한 적이 있습니다.
